conn/sticky_linux.go

```diff
@@ -0,0 +1,121 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2022 WireGuard LLC. All Rights Reserved.
+ */
+
+package conn
+
+import (
+	"net/netip"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// TODO: macOS, FreeBSD and other BSDs likely do support this feature set, but
+// use alternatively named flags and need ports and require testing.
+
+// getSrcFromControl parses the control for PKTINFO and if found updates ep with
+// the source information found.
+func getSrcFromControl(control []byte, ep *StdNetEndpoint) {
+	ep.ClearSrc()
+
+	var (
+		hdr  unix.Cmsghdr
+		data []byte
+		rem  []byte = control
+		err  error
+	)
+
+	for len(rem) > unix.SizeofCmsghdr {
+		hdr, data, rem, err = unix.ParseOneSocketControlMessage(control)
+		if err != nil {
+			return
+		}
+
+		if hdr.Level == unix.IPPROTO_IP &&
+			hdr.Type == unix.IP_PKTINFO {
+
+			info := bufCast[unix.Inet4Pktinfo](data)
+			ep.src.Addr = netip.AddrFrom4(info.Spec_dst)
+			ep.src.ifidx = info.Ifindex
+
+			return
+		}
+
+		if hdr.Level == unix.IPPROTO_IPV6 &&
+			hdr.Type == unix.IPV6_PKTINFO {
+
+			info := bufCast[unix.Inet6Pktinfo](data)
+			ep.src.Addr = netip.AddrFrom16(info.Addr)
+			ep.src.ifidx = int32(info.Ifindex)
+
+			return
+		}
+	}
+}
+
+// cast a buffer to a type directly, only for types known to be safe to do so.
+// panics if buf is of insufficient size.
+func bufCast[T unix.Inet4Pktinfo | unix.Inet6Pktinfo](buf []byte) (t T) {
+	if len(buf) < int(unsafe.Sizeof(t)) {
+		panic("byteCast: buffer too small")
+	}
+	return *(*T)(unsafe.Pointer(&buf[0]))
+}
+
+// setSrcControl parses the control for PKTINFO and if found updates ep with
+// the source information found.
+func setSrcControl(control *[]byte, ep *StdNetEndpoint) {
+	*control = (*control)[:cap(*control)]
+	if len(*control) < int(unsafe.Sizeof(unix.Cmsghdr{})) {
+		*control = (*control)[:0]
+		return
+	}
+
+	if ep.src.ifidx == 0 && !ep.SrcIP().IsValid() {
+		*control = (*control)[:0]
+		return
+	}
+
+	if len(*control) < srcControlSize {
+		*control = (*control)[:0]
+		return
+	}
+
+	hdr := (*unix.Cmsghdr)(unsafe.Pointer(&(*control)[0]))
+	if ep.SrcIP().Is4() {
+		hdr.Level = unix.IPPROTO_IP
+		hdr.Type = unix.IP_PKTINFO
+		hdr.SetLen(unix.CmsgLen(unix.SizeofInet4Pktinfo))
+
+		info := (*unix.Inet4Pktinfo)(unsafe.Pointer(&(*control)[unix.SizeofCmsghdr]))
+		info.Ifindex = ep.src.ifidx
+		if ep.SrcIP().IsValid() {
+			info.Spec_dst = ep.SrcIP().As4()
+		}
+	} else {
+		hdr.Level = unix.IPPROTO_IPV6
+		hdr.Type = unix.IPV6_PKTINFO
+		hdr.Len = unix.SizeofCmsghdr + unix.SizeofInet6Pktinfo
+
+		info := (*unix.Inet6Pktinfo)(unsafe.Pointer(&(*control)[unix.SizeofCmsghdr]))
+		info.Ifindex = uint32(ep.src.ifidx)
```

Comment:
Did you test that this actually works in the absence of the v6 scope ID? Or is this code also preserving v6 scope id in the sockaddr? https://git.zx2c4.com/wireguard-tools/tree/contrib/sticky-sockets/sticky-sockets.c appears to preserve the scope id in the sockaddr, which is why I'm wondering.

More generally, do sticky sockets work? Did you test?

Reply:
getSrcFromControl records an in6_addr and ifindex from an ipv6_pktinfo, that is exactly all of the information contained when the packets are received, there is no sockaddr form of source address from this API.

setSrcControl does precisely the inverse, it writes the in6_addr and ifindex into the control mesasage.

sticky-sockets.c is (re)using a struct sockaddr to get a slot to store the ifindex, but what it is writing into the sin6_scope_id field is ifindex from the pktinfo (https://git.zx2c4.com/wireguard-tools/tree/contrib/sticky-sockets/sticky-sockets.c#n160), and it restores it on send here: https://git.zx2c4.com/wireguard-tools/tree/contrib/sticky-sockets/sticky-sockets.c#n85. The behavior of these two implementations is the same - the sticky-sockets.c is using a field of a different name for shuttling the same data - here the name matches the name in the data source and target API.

There is also automated test coverage added for these behaviors here: https://github.com/WireGuard/wireguard-go/pull/64/files/798bed0b5d9283a45d9f3e78da7f5acd7477b7a6#diff-0fd41a3f9a11bd85ee1bafb4f333c956cfd7198806afbf1fc963bc5411ca9289R79 and in additional cases in that file.

Reply:
Super, thanks.